crates/rover-std/src/fs.rs

```diff
@@ -54,23 +55,70 @@ impl Fs {
         C: AsRef<[u8]>,
     {
         let path = path.as_ref();
-        let contents = str::from_utf8(contents.as_ref()).with_context(|| {
-            format!(
-                "tried to write contents to {} that was invalid UTF-8",
-                &path
-            )
-        })?;
-        if !path.exists() {
-            File::create(path)
-                .with_context(|| format!("{} does not exist and it could not be created", &path))?;
-        }
-        if !path.exists() {
-            File::create(path)
-                .with_context(|| format!("{} does not exist and it could not be created", &path))?;
+        tracing::info!("checking existence of parent path in '{}'", path);
+        // Grab the parent path then attempt to canonicalize it, we can't just canonicalize the
+        // entire path because that would entail the file existing, which of course it doesn't yet.
+        let mut canonical_final_path = if let Some(parent_path) = path.parent() {
+            tracing::debug!("attempting to canonicalize parent path '{}'", parent_path);
+            Self::upsert_path_exists(parent_path)
+        } else {
+            // If the has no parent, the only conceivable case is that we are looking at the root,
+            // and as that cannot be a file we should bail out at this point
+            return Err(anyhow::anyhow!("cannot write file to root or prefix {}", path).into());
+        }?;
+
+        return if let Some(file_name) = path.file_name() {
+            // Create the final version of the path we want to create
+            canonical_final_path.push(file_name);
+            tracing::debug!("final canonical path is {}", canonical_final_path);
+            // Setup a file pointer
+            let mut file = OpenOptions::new()
+                .write(true)
+                .create(true)
+                .truncate(true)
+                .open(path)
+                .with_context(|| {
+                    format!(
+                        "tried to open {} but was unable to do so",
+                        &canonical_final_path
+                    )
+                })?;
+            tracing::info!("writing {} to disk", &canonical_final_path);
+            // Actually write the file out to where it needs to be
+            file.write(contents.as_ref())
+                .with_context(|| format!("could not write {}", &canonical_final_path))?;
+            Ok(())
+        } else {
+            // If we can't pull out a final element, then we're in a weird case but should bail
+            // out as there's nothing sensible we can do.
+            Err(anyhow::anyhow!("cannot write to a path without a final element {}", path).into())
+        };
+    }
+
+    /// Given a path, where some elements may not exist, it will return the canonical
+    /// representation of the path, AND create any missing interim directories.
+    fn upsert_path_exists(path: &Utf8Path) -> Result<Utf8PathBuf, anyhow::Error> {
```

Comment:
does this need to be a method off of `Fs`? trying to figure out if it adds anything semantically to have it as part of `Fs` or if folks will get confused seeing it there

Reply:
I guess there's a precedent with `create_dir_all`, but I sort of think _that_ fn is confusing to have under `Fs` too, but I guess it gets used throughout the codebase; makes me glad that `upsert_path_exists` doesn't need to be marked `pub`

Reply:
I feel like since it's a private method that's kind of ok? You could in-line it, but it felt like the method it was part of was complex enough without adding a big closure in the middle. Happy to change it though if you think that's better?

Reply:
I really don't know! I don't think it _really_ matters either way; it's also worth calling out that I don't know much about the conventions rover has been using in terms of code organization